ismatec/util.py

```diff
@@ -159,156 +140,115 @@ def close(self):
-                logger.debug(f'received message: "{line}"')
+        return msg.strip()
 
     def close(self):
         """Release resources."""
         self.socket.shutdown(socket.SHUT_RDWR)
         self.socket.close()
 
 
-class Protocol:
-    """Convert to various (dumb) datatypes used for the protocol."""
-
-    def _time1(self, number, units='s'):
-        """Convert number to 'time type 1'.
-
-        1-8 digits, 0 to 35964000 in units of 0.1s (0 to 999 hr)
-        """
-        unit_options = {
-            's': 10,
-            'm': 600,
-            'h': 36000,
-        }
-        number = int(number * unit_options[units])
-        return str(min(number, 35964000)).replace('.', '')
-
-    def _time2(self, number, units='s'):
-        """Convert number to 'time type 2'.
-
-        This is an 8-digit left-padded version of `_time1`.
-        """
-        return self._time1(number, units).zfill(8)
-
-    def _volume1(self, number):
-        """Convert number to 'volume type 1'.
-
-        mmmmEse — Represents the scientific notation of m.mmm x 10se.
-        For example, 1.200 x 10-2 is represented with 1200E-2. Note
-        that the decimal point is inferred after the first character.
-        """
-        s = f'{abs(number):.3e}'
-        return f'{s[0]}{s[2:5]}E{s[-3]}{s[-1]}'
-
-    def _volume2(self, number):
-        """Convert number to 'volume type 2'.
-
-        This is undocumented. It appears to be 'volume type 1' without
-        the E, and can be mL or mL/min depending on use case.
-        """
-        return self._volume1(number).replace('E', '')
-
-    def _discrete2(self, number):
-        """Convert number to 'discrete type 2'.
-
-        Four characters representing a discrete integer value in base
-        10. The value is right-justified. Unused digits to the left are
-        zeros.
-        """
-        s = str(number).strip('0')
-        whole, decimals = s.split('.')
-        return '%04d' % int(whole + decimals)
-
-    def _discrete3(self, number):
-        """Convert number to 'discrete type 3'.
-
-        Six characters in base 10. The value is right-justified.
-        Unused digits to the left are zeros.
-        """
-        assert 0 <= number < 1_000_000
-        return str(number).zfill(6)
-
-    class Mode(Enum):
-        """Possible operating modes."""
-
-        RPM = 'L'
-        FLOWRATE = 'M'
-        VOL_AT_RATE = 'O'
-        VOL_OVER_TIME = 'G'
-        VOL_PAUSE = 'Q'
-        TIME = 'N'
-        TIME_PAUSE = 'P'
-
-    class Setpoint(Enum):
-        """Possible setpoint types."""
-
-        RPM = '0'
-        FLOWRATE = '1'
-
-    class Rotation(Enum):
-        """Possible rotation directions."""
-
-        CLOCKWISE = 'J'
-        COUNTERCLOCKWISE = 'K'
-
-    Tubing = [
-        0.13, 0.19, 0.25, 0.38, 0.44, 0.51, 0.57, 0.64, 0.76, 0.89, 0.95, 1.02, 1.09,
-        1.14, 1.22, 1.30, 1.43, 1.52, 1.65, 1.75, 1.85, 2.06, 2.29, 2.54, 2.79, 3.17
-    ]
+class Mode(Enum):
```

Comment:
All this stuff feels like it's hiding when in the class. Like the class was just a holding container for it.

Pulling it up a layer makes both this file and the main driver cleaner.

Also, pulling it up allowed the linter to identify unused methods (ex. time1) and variables (tubing) that I then deleted.

Reply:
Time Type 1 is used by methods that aren't implemented yet.
<img width="873" alt="Screenshot 2023-03-03 at 4 40 21 PM" src="https://user-images.githubusercontent.com/52292902/222845101-c10ba6e3-a36b-4009-b51a-fdd6f962fd2a.png">


And the list of tubing sizes should probably be used to enforce that only those particular sizes are set.  Presumably others will fail.

Reply:
`Protocol` was shared across the main driver and the mock, so pulling it up into the driver killed the mock.  I'll see if it's possible for the mock to inherit/override.